File: src/solar/solar-common/common/SolarTrigger.ts
/** @noSelfInFile */
import ArrayUtil from "@/ArrayUtil";
import ErrorMsgHelper from "@/ErrorMsgHelper";


declare global {
    type SolarTriggerArray = SolarTrigger[];
}

let index = 0;
export default class SolarTrigger implements Destroyable {
    private static _sl_TriggeringSolarTrigger: SolarTrigger = null;
    /** uuid是异步的 可能ui函数也在注册SolarTrigger */
    readonly uuid: string = "stri" + index++;
    enable: boolean = true;
    onDestroy: ((this: void, solarTrigger: SolarTrigger) => void) = null;
    actionFuncs: (((this: void, solarTrigger: SolarTrigger, data?: any) => void)[]) = [];
    private solarTriggers: SolarTriggerArray = null;
    private _sl_isDestroyed = false;

    /**
     *
     * @param actionFunc 触发时的回调函数
     * @param solarTriggers 触发器寄生的触发数组
     */
    constructor(actionFunc: (this: void, solarTrigger: SolarTrigger, data?: any) => void, solarTriggers: SolarTriggerArray) {
        this.addAction(actionFunc);
        this.solarTriggers = solarTriggers;
        this.solarTriggers.push(this);
    }

    public addAction(actionFunc: (this: void, solarTrigger: SolarTrigger, data?: any) => void) {
        this.actionFuncs.push(actionFunc);
    }

    /**
     * 移除触发器动作
     */
    public removeAction(actionFunc: (this: void, solarTrigger: SolarTrigger, data?: any) => void) {
        this.actionFuncs.splice(this.actionFuncs.indexOf(actionFunc), 1);
    }

    /**
     * 清空触发器动作
     */
    public removeActions() {
        this.actionFuncs = [];
    }


    public exec(data?: any) {
        SolarTrigger._sl_TriggeringSolarTrigger = this;
        if (this._sl_isDestroyed) {
            return
        }
        if (this.enable == false) {
            return
        }
        //即使报错也不要影响后续的逻辑 以尽可能提高游戏体验
        xpcall(() => {
            for (let actionFunc of this.actionFuncs) {
                actionFunc(this, data);
            }
        }, ErrorMsgHelper.error_handle);
        SolarTrigger._sl_TriggeringSolarTrigger = null;
    }


    public destroy() {
        this._sl_isDestroyed = true;
        this.onDestroy?.(this);
        this.actionFuncs = null;
        ArrayUtil.removeElement(this.solarTriggers, this)
        this.solarTriggers = null;
    }

    /**
     * 在SolarTrigger动作事件中获得 当前的触发器
     */
    static getTriggeringSolarTrigger(): SolarTrigger {
        return SolarTrigger._sl_TriggeringSolarTrigger;
    }
}

File: scripts/solar/solar-common/common/SolarTrigger.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 2,["9"] = 2,["10"] = 3,["11"] = 3,["12"] = 10,["13"] = 11,["14"] = 11,["15"] = 11,["16"] = 26,["17"] = 14,["18"] = 14,["19"] = 14,["20"] = 15,["21"] = 16,["22"] = 17,["23"] = 18,["24"] = 19,["25"] = 27,["26"] = 28,["27"] = 29,["28"] = 29,["29"] = 26,["30"] = 32,["31"] = 33,["32"] = 33,["33"] = 32,["34"] = 39,["35"] = 40,["36"] = 40,["37"] = 40,["38"] = 40,["39"] = 40,["40"] = 39,["41"] = 46,["42"] = 47,["43"] = 46,["44"] = 51,["45"] = 52,["46"] = 53,["49"] = 56,["52"] = 60,["53"] = 60,["54"] = 61,["55"] = 62,["57"] = 60,["58"] = 60,["59"] = 60,["60"] = 65,["61"] = 51,["62"] = 69,["63"] = 70,["64"] = 71,["66"] = 71,["68"] = 72,["69"] = 73,["70"] = 74,["71"] = 69,["72"] = 80,["73"] = 81,["74"] = 80,["75"] = 12});
local ____exports = {}
local ____ArrayUtil = require("solar.solar-common.util.lang.ArrayUtil")
local ArrayUtil = ____ArrayUtil.default
local ____ErrorMsgHelper = require("solar.solar-common.common.ErrorMsgHelper")
local ErrorMsgHelper = ____ErrorMsgHelper.default
local index = 0
____exports.default = __TS__Class()
local SolarTrigger = ____exports.default
SolarTrigger.name = "SolarTrigger"
function SolarTrigger.prototype.____constructor(self, actionFunc, solarTriggers)
    local ____index_1 = index
    index = ____index_1 + 1
    self.uuid = "stri" .. tostring(____index_1)
    self.enable = true
    self.onDestroy = nil
    self.actionFuncs = {}
    self.solarTriggers = nil
    self._sl_isDestroyed = false
    self:addAction(actionFunc)
    self.solarTriggers = solarTriggers
    local ____self_solarTriggers_0 = self.solarTriggers
    ____self_solarTriggers_0[#____self_solarTriggers_0 + 1] = self
end
function SolarTrigger.prototype.addAction(self, actionFunc)
    local ____self_actionFuncs_2 = self.actionFuncs
    ____self_actionFuncs_2[#____self_actionFuncs_2 + 1] = actionFunc
end
function SolarTrigger.prototype.removeAction(self, actionFunc)
    __TS__ArraySplice(
        self.actionFuncs,
        __TS__ArrayIndexOf(self.actionFuncs, actionFunc),
        1
    )
end
function SolarTrigger.prototype.removeActions(self)
    self.actionFuncs = {}
end
function SolarTrigger.prototype.exec(self, data)
    ____exports.default._sl_TriggeringSolarTrigger = self
    if self._sl_isDestroyed then
        return
    end
    if self.enable == false then
        return
    end
    xpcall(
        function()
            for ____, actionFunc in ipairs(self.actionFuncs) do
                actionFunc(self, data)
            end
        end,
        ErrorMsgHelper.error_handle
    )
    ____exports.default._sl_TriggeringSolarTrigger = nil
end
function SolarTrigger.prototype.destroy(self)
    self._sl_isDestroyed = true
    local ____opt_3 = self.onDestroy
    if ____opt_3 ~= nil then
        ____opt_3(self)
    end
    self.actionFuncs = nil
    ArrayUtil:removeElement(self.solarTriggers, self)
    self.solarTriggers = nil
end
function SolarTrigger.getTriggeringSolarTrigger(self)
    return ____exports.default._sl_TriggeringSolarTrigger
end
SolarTrigger._sl_TriggeringSolarTrigger = nil
return ____exports
